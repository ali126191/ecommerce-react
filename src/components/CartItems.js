import React from 'react';
import { connect } from 'react-redux';
import { updateItemQuantity, removeCartItem } from '../actions/cart';

export const CartItems = (props) => (
    <div>
        {props.cart.map((phone, index) => (
            <div key={phone.id}>
                {phone.name}-
                {phone.quantity}
                <button 
                    onClick={() => {
                        props.dispatch(updateItemQuantity(phone.id, { quantity: 1 }))
                    }}
                >
                    Add quantity
                </button>
                <button 
                    onClick={() => {
                        if (phone.quantity > 0) {
                            props.dispatch(updateItemQuantity(phone.id, { quantity: -1 }))
                        }
                    }}
                >
                     Decrease quantity
                </button>
                <button onClick={() => {
                    props.dispatch(removeCartItem(phone.id))
                }}>
                    Delete
                </button>
            </div>
        ))}
        
    </div>
)

const mapStateToProps = (state) => {
    function getUnique(arr, comp) {
        const unique = arr.map(e => e[comp])
           // store the keys of the unique objects
        .map((e, i, final) => final.indexOf(e) === i && i)
        // eliminate the dead keys & store unique objects
        .filter(e => arr[e]).map(e => arr[e]);
        
         return unique;
      }

    return {
        cart: getUnique(state.cart, 'id')
    }
}

export default connect(mapStateToProps)(CartItems);