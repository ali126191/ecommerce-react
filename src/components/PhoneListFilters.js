import React from 'react';
import { connect } from 'react-redux';
import { setTextFilter, sortByPrice, sortByRandom } from '../actions/filters';

export class PhoneListFilters extends React.Component {
    onTextChange = (e) => {
        this.props.setTextFilter(e.target.value);
     }
     onSortChange = (e) => {
        if (e.target.value === 'price') {
            this.props.sortByPrice();
        } else if (e.target.value === 'random') {
            this.props.sortByRandom();
        }
    }

    render() {
        return (
            <div>
                <input 
                    type="text" 
                    value={this.props.filters.text}
                    onChange={this.onTextChange}  
                />
                <select 
                    value={this.props.filters.sortBy}
                    onChange={this.onSortChange}
                >
                     <option value="price">Price</option>
                     <option value="random">Random</option>
                </select>
            </div>
        )
    }
}

const mapStateToProps = (state) => {
    return {
        filters: state.filters
    };
};

const mapDispatchToProps = (dispatch) => ({
    setTextFilter: (text) => dispatch(setTextFilter(text)),
    sortByPrice: () => dispatch(sortByPrice()),
    sortByRandom: () => dispatch(sortByRandom())
});

export default connect(mapStateToProps, mapDispatchToProps)(PhoneListFilters);