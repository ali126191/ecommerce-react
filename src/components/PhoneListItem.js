import React from 'react';
import { connect } from 'react-redux';
import { addToCart } from '../actions/cart';

export const PhoneListItem = ({ name, price, quantity, id, dispatch }, props) => (
    <div>
        {name}: 
        {price}, 
        {quantity}-
        <button onClick={(e) => {
                dispatch(addToCart({ name, price, quantity, id }))
            }}
        >
          Add To Cart
        </button>
    </div>
)

const mapToState = (state) => {
    return {
        cart: state.cart
    }
}

export default connect(mapToState)(PhoneListItem);