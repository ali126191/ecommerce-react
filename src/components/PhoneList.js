import React from 'react';
import { connect } from 'react-redux';
import PhoneListItem from './PhoneListItem';
import SelectExpenses from '../selectors/cart';

export const PhoneList = (props) => (
    <div>
        <h1>Phone List</h1>
        {props.phones.map((phone) => (
            <PhoneListItem key={phone.id} {...phone}/>
        ))}
    </div>
)

const mapStateToProps = (state) => {

    function getUnique(arr, comp) {
        const unique = arr.map(e => e[comp])
           // store the keys of the unique objects
        .map((e, i, final) => final.indexOf(e) === i && i)
        // eliminate the dead keys & store unique objects
        .filter(e => arr[e]).map(e => arr[e]);
        
         return unique;
      }

    return {
        phones: SelectExpenses(state.phones, state.filters),
        cart: getUnique(state.cart, 'id'),
        filters: state.filters
    }
}

export default connect(mapStateToProps)(PhoneList)