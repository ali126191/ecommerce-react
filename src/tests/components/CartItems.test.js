import React from 'react';
import { shallow } from 'enzyme';
import toJSON from 'enzyme-to-json';
import { CartItems } from '../../components/CartItems';
import phones from '../fixtures/phones';

test('should render CartItems correctly', () => {
    const wrapper = shallow(<CartItems cart={phones} />);
    expect(toJSON(wrapper)).toMatchSnapshot();
})