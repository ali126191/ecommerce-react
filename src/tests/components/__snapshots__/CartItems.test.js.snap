// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render CartItems correctly 1`] = `
<div>
  <div
    key="1"
  >
    Google Pixel
    -
    0
    <button
      onClick={[Function]}
    >
      Add quantity
    </button>
    <button
      onClick={[Function]}
    >
      Decrease quantity
    </button>
    <button
      onClick={[Function]}
    >
      Delete
    </button>
  </div>
  <div
    key="2"
  >
    Black Berry
    -
    0
    <button
      onClick={[Function]}
    >
      Add quantity
    </button>
    <button
      onClick={[Function]}
    >
      Decrease quantity
    </button>
    <button
      onClick={[Function]}
    >
      Delete
    </button>
  </div>
  <div
    key="3u"
  >
    Iphone xs
    -
    0
    <button
      onClick={[Function]}
    >
      Add quantity
    </button>
    <button
      onClick={[Function]}
    >
      Decrease quantity
    </button>
    <button
      onClick={[Function]}
    >
      Delete
    </button>
  </div>
</div>
`;
